File: Balancing_selection/MuteBass/Old/Find_Alternative_ancestral.jl
using DataFrames, CSV

# Capture command-line arguments
function main()
    args = ARGS
    if length(args) < 2
        println("Usage: julia script.jl species_list path_to_data")
        return
    end
    
    # Parse species list and data path from command-line arguments
    species_list = split(args[1], ",")
    path_2_data = args[2]
    
    # Combine data for all species
    Merged_dt = Combine_data(species_list, path_2_data)
    
    # Determine the majority ancestral state
    Alternative_ancestral_state = Get_majority(Merged_dt)
    
    # Set alternative ancestral alleles where 'nan' exists
    Alternative_ancestral_output = Set_alternative_ancestral_nan(path_2_data, Alternative_ancestral_state)
    
    # Print final output
    println(Alternative_ancestral_state)
end

# Function to combine data from species-specific files
function Combine_data(species_list, path_2_data)
    all_species_data = DataFrame(Chromosome = String[], Position = Int[], 
                                  Allele1_frequency = String[], Allele2_frequency = String[], 
                                  Allele1 = String[], frequency1 = Float64[], 
                                  Allele2 = String[], frequency2 = Float64[], 
                                  Species = String[])  # Initialize with specific columns
    
    for species in species_list
        # Read each species-specific file with .frq extension
        species_files = filter(x -> occursin(r"^$species.*\.frq$", x), readdir(path_2_data))
        
        for file in species_files
            # Read file and assign column names manually
            file_path = joinpath(path_2_data, file)
            frequencies = CSV.read(file_path, DataFrame, header = false)
            rename!(frequencies, [:Chromosome, :Position, :V3, :V4, :Allele1_frequency, :Allele2_frequency])
            
            # Split allele frequency columns (Allele1:frequency1, Allele2:frequency2)
            frequencies[!, :Allele1] = first.(split.(frequencies[!, :Allele1_frequency], ":"))
            frequencies[!, :frequency1] = parse.(Float64, last.(split.(frequencies[!, :Allele1_frequency], ":")))
            frequencies[!, :Allele2] = first.(split.(frequencies[!, :Allele2_frequency], ":"))
            frequencies[!, :frequency2] = parse.(Float64, last.(split.(frequencies[!, :Allele2_frequency], ":")))
            
            # Remove unnecessary columns and add species identifier
            select!(frequencies, Not([:V3, :V4]))
            frequencies[!, :Species] .= species
            
            # Filter out rows where frequency1 is NaN
            filter!(row -> !isnan(row.frequency1), frequencies)
            
            # Append data to all_species_data
            append!(all_species_data, frequencies)  # Combine into DataFrame
        end
    end
    
    return all_species_data  # Return combined DataFrame
end

# Function to get the majority allele by summing frequencies for each SNP
function Get_majority(complete_df::DataFrame)
    SNPs_info = DataFrame(Position = String[], MajorityAllele = String[])
    
    for pos in unique(complete_df.Position)
        current_SNP = complete_df[complete_df.Position .== pos, :]
        allele1_sum = sum(current_SNP.frequency1)
        allele2_sum = sum(current_SNP.frequency2)
        
        majority_allele = allele1_sum > allele2_sum ? unique(current_SNP.Allele1)[1] : unique(current_SNP.Allele2)[1]
        
        push!(SNPs_info, (Position = string(pos), MajorityAllele = majority_allele))
    end
    
    return SNPs_info
end

# Function to set alternative ancestral alleles where 'nan' exists
function Set_alternative_ancestral_nan(path_2_data, Alternative_ancestral_state)
    ancestral_file = joinpath(path_2_data, "ancestral.alleles")
    current_ancestral = CSV.read(ancestral_file, DataFrame, header = false)
    rename!(current_ancestral, [:Position, :AncestralAllele])
    
    for row in eachrow(current_ancestral)
        if row.AncestralAllele == "nan"
            alternative_state = Alternative_ancestral_state[Alternative_ancestral_state.Position .== row.Position, :MajorityAllele]
            if !isempty(alternative_state)
                row.AncestralAllele = alternative_state[1]  # Update with found allele
            end
        end
    end
    
    # Return the updated DataFrame with the new ancestral alleles
    return current_ancestral
end

# Run the main function
main()

File: Introgression/Relate/Find_Alternative_ancestral.r
# Load the 'tidyr' library for data manipulation functions
library(tidyr)

# 'args' is used to capture command-line arguments. 
# The first argument is a list of species separated by commas, which is split into a vector.
# The second argument is the path to the data files.
args <- commandArgs(trailingOnly = TRUE)
species_list <- strsplit(args[1], ",")[[1]]  # Split species into a list
path_2_data <- args[2]  # Store the path to data

# Function to combine data from species-specific files
Combine_data <- function(species_list, path_2_data) {
  list = list()  # Initialize an empty list to store data from different species
  sp = 1  # Index for the list
  
  # Loop through each species in the species_list
  for (species in species_list) {
    # Find the species-specific file(s) with ".frq" extension and store the full file path
    species_frequencies_file <- list.files(path = path_2_data, pattern = paste0("^", species, ".*\\.frq$"), full.names = TRUE)
    # Read the file into a dataframe
    frequencies <- read.table(species_frequencies_file)
    # Separate the 'V5' column (e.g., Allele1:frequency1) into two columns: Allele1 and frequency1
    frequencies <- separate(frequencies, V5, into = c("Allele1", "frequency1"), sep = ":")
    # Separate the 'V6' column (e.g., Allele2:frequency2) into two columns: Allele2 and frequency2
    frequencies <- separate(frequencies, V6, into = c("Allele2", "frequency2"), sep = ":")
    # Add a new column to track the species name
    frequencies$Species = species
    # Remove the 'V3' and 'V4' columns from the dataframe as they are not needed
    frequencies <- frequencies[,-c(3,4)]
    # Filter out rows where 'frequency1' has the value '-nan'
    frequencies <- frequencies[frequencies$frequency1 != "-nan",]
    # Store the cleaned data for this species in the list
    list[[sp]] = frequencies
    sp = sp + 1  # Increment the index
  }
  
  # Combine all dataframes in the list into one large dataframe using rbind (row binding)
  complete_df <- as.data.frame(do.call(rbind, list))
  
  # Return the combined dataframe
  return(complete_df)
}

# Function to determine the majority allele based on frequency for each SNP
Get_majority <- function(complete_df) {
  list = list()  # Initialize an empty list to store majority allele information
  SNP = 1  # Index for the list
  
  # Loop through each unique SNP position (V2 column in complete_df)
  for (i in unique(complete_df$V2)) {  
    # Extract all rows corresponding to the current SNP position
    current_SNP <- complete_df[complete_df$V2 == i,]
    # Calculate the sum of frequencies for Allele1 and Allele2 (ignore NA values)
    weight_allele1 <- sum(as.numeric(current_SNP$frequency1), na.rm = TRUE)
    weight_allele2 <- sum(as.numeric(current_SNP$frequency2), na.rm = TRUE)
    # If both allele frequencies are equal, randomly pick one of the alleles
    if (weight_allele1 == weight_allele2) {
      result = c(i, sample(c(unique(current_SNP$Allele1), unique(current_SNP$Allele2)), 1))
    } else {
      # Otherwise, choose the allele with the higher total frequency
      result <- c(i, ifelse(weight_allele1 > weight_allele2, unique(current_SNP$Allele1), unique(current_SNP$Allele2)))
    }
    
    # Store the SNP position and the majority allele in the list
    list[[SNP]] <- result
    SNP = SNP + 1  # Increment the index
  }
  # Combine all results into a dataframe
  SNPs_info <- as.data.frame(do.call(rbind, list))
  # Return the dataframe containing SNP positions and majority alleles
  return(SNPs_info)
}

Set_alternative_ancestral_nan <- function(path_2_data, Alternative_ancestral_state) {
    current_ancestral <- read.table(paste(path_2_data, "/ancestral.alleles", sep=""))
    nan_SNPs <- c()  # Initialize a vector to store SNPs with 'nan'
    
    for (i in 1:nrow(current_ancestral)) {
        current_position <- current_ancestral[i,]
        
        # Check if the ancestral allele is 'nan'
        if (current_position[,2] == "nan") {
            current_SNP <- current_position[,1]
            
            # Retrieve the alternative allele for the current SNP
            alternative_state <- Alternative_ancestral_state[Alternative_ancestral_state$V1 == current_SNP, 2]
            
            # If alternative state is not available, add the SNP to nan_SNPs
            if (length(alternative_state) == 0 || is.na(alternative_state)) {
                current_ancestral[i, 2] <- "nan"
                nan_SNPs <- c(nan_SNPs, current_SNP)  # Append SNP with nan
            } else {
                current_ancestral[i, 2] <- alternative_state
            }
        }
    }
    
    # Return both the updated ancestral alleles and the list of SNPs with 'nan'
    return(list(updated_ancestral = current_ancestral, nan_SNPs = nan_SNPs))
}

State_absent_from_all_taxa_used_4_ancestral <- function(hap_file,Alternative_ancestral_state){
	hap_file <- read.table(hap_file)
	hap_file <- hap_file[,c(3,4,5)]
	Final_Alternative_ancestral_state <- Alternative_ancestral_state$updated_ancestral
	for (i in Alternative_ancestral_state$nan_SNPs){
		Allele1 <- hap_file[hap_file$V3==i,2]
		Allele2 <- hap_file[hap_file$V3==i,3]
		random_Allele <- sample(c(Allele1,Allele2),1)
		Final_Alternative_ancestral_state[Final_Alternative_ancestral_state$V1==i,2] = random_Allele
	}
return(Final_Alternative_ancestral_state[,2])
}

# Combine data for all species
Merged_dt <- Combine_data(species_list, path_2_data)

# Determine the majority ancestral state
Alternative_ncestral_state <- Get_majority(Merged_dt)

# Set alternative ancestral alleles where 'nan' exists
Alternative_ancestral_state <- Set_alternative_ancestral_nan(path_2_data, Alternative_ncestral_state)

# Handle the case were all individuals used as ancestral state are missing data by picking one of the two SNPs from the hap file randomly
Final_ancestral_state <- State_absent_from_all_taxa_used_4_ancestral(args[3],Alternative_ancestral_state)

# Print the new ancestral.state file 
write.table(x = Final_ancestral_state, file = paste(path_2_data,"/ancestral.alleles",sep=""),quote=FALSE,row.names=FALSE, col.names=FALSE)

